packages/connection-storage/src/connection-provider.ts

```diff
@@ -0,0 +1,90 @@
+import { ConnectionInfo } from '@mongodb-js/connection-info';
+import { ConnectionStorage } from './connection-storage';
+import ConnectionString from 'mongodb-connection-string-url';
+import { merge } from 'lodash';
+
+export type ConnectionStatus =
+  | 'connected'
+  | 'disconnected'
+  | 'connecting'
+  | 'failed';
+export type StatusAwareConnectionInfo = Omit<
+  ConnectionInfo & {
+    status: ConnectionStatus;
+  },
+  'favorite'
+>;
+
+/**
+ * Exposes an interface that will be implemented for each deployment environment to
+ * get access to the connections on that environment. For Compass Desktop, it will
+ * rely on the ConnectionStorage, for the Data Explorer, on the Atlas API (or another
+ * facade on top of that).
+ */
+export interface ConnectionProvider {
```

Comment:
Isn't this then a`ConnectionStorageProvider` rather than a `ConnectionProvider?

Reply:
Well, not exactly. It provides the list of connections. In desktop it uses the ConnectionStorage because that's how it is already implemented and I didn't want to rewrite everything. In web, it would use likely the Atlas API, so it won't depend on any storage provider.

Reply:
Ahh, now i understand this is not just an interface for storages to implement, is a new pair of classes independent from connection storage.

Sorry for being a pain here, I'm thinking about someone joining later on, without context, and having to districate these concepts. The desing here is pretty clear to me, but i have an almost perfect amount of background. I have the feeling that we could just keep `ConnectionStorage` here as interface and just rename the current "ConnectionStorage" to `CompassConnectionStorage`, is there a downside?

The additional sorting can be sent inside the `CompassConnectionStorage` if is something that needs to be implemented in the API of de, or just be part of the stores as functions, if is more of a presentational detail, and we don't need the API to return them in that way.

Reply:
Yeah, that's something that I considered also at the beginning, but this approach is a bit more complex because the ConnectionStorage is a bit more convoluted to refactor because it has been designed with the flow of the desktop application in mind (which makes sense to be fair). 

For example, it has two implementations that share the same interface: one for the renderer process and another one for the backend process. Also, the ConnectionStorage expose things that won't make sense in the Data Explorer:

```
| 'loadAll'
    | 'load'
    | 'getLegacyConnections' // < this
    | 'save'
    | 'delete'
    | 'deserializeConnections' // < this
    | 'importConnections' // < this
    | 'exportConnections' // < this 
```

Also, it's implemented as a static class (it's a singleton), so while it's possible to do some dependency injection here, it would be more complicated and would probably patch the typescript type system to make it work without random castings.

You can see how the singleton is defined in `packages/connection-storage/src/renderer.ts`.
And the backend implementation is: `packages/connection-storage/src/connection-storage.ts`.

So it adds a new layer only for the Compass implementation and adds some safe guards (validations that were on the UI only for example are now in the entry point of working with connections).

Does that sound good?

Reply:
Yeah i see the struggle a bit better now, i think is better to "adapt" connection storage to be a normal class (and a singleton),  and avoid the additional pair of classes. Duck typing here is is enough to ignore the unnecessary methods in the shared areas of Compass.